lib/Dialect/HW/Transforms/FlattenIO.cpp

```diff
@@ -0,0 +1,292 @@
+//===- FlattenIO.cpp - HW I/O flattening pass -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWPasses.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt;
+
+static bool isStructType(Type type) {
+  return hw::getCanonicalType(type).isa<hw::StructType>();
+}
+
+// Legal if no in- or output type is a struct.
+static bool isLegalFuncLikeOp(FunctionOpInterface moduleLikeOp) {
+  bool legalResults =
+      llvm::none_of(moduleLikeOp.getResultTypes(), isStructType);
+  bool legalArgs = llvm::none_of(moduleLikeOp.getArgumentTypes(), isStructType);
+
+  return legalResults && legalArgs;
+}
+
+static llvm::SmallVector<Type> getInnerTypes(hw::StructType t) {
+  llvm::SmallVector<Type> inner;
+  t.getInnerTypes(inner);
+  return inner;
+}
+
+namespace {
+
+// Replaces an output op with a new output with flattened (exploded) structs.
+struct OutputOpConversion : public OpConversionPattern<hw::OutputOp> {
+  OutputOpConversion(TypeConverter &typeConverter, MLIRContext *context,
+                     DenseSet<Operation *> *outputOpVisited)
+      : OpConversionPattern(typeConverter, context),
+        outputOpVisited(outputOpVisited) {}
+
+  LogicalResult
+  matchAndRewrite(hw::OutputOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<Value> convOperands;
+
+    // Flatten the operands.
+    for (auto operand : adaptor.getOperands()) {
+      if (auto structType = operand.getType().dyn_cast<hw::StructType>()) {
```

Comment:
Is it ok to not use `hw::getCanonicalType` here?

Reply:
Good catch - alias types will anyways be elided through this pass, but I'll add a test just to exercise the path.